File: ghost_healer/core/credentials.py
"""
Credential store — install-only SDK access + optional user override.

Built-in public key is bundled with the SDK; no manual login required for hosted Brain.
"""
import json
import os
from pathlib import Path
from typing import Any, Dict, Optional

DEFAULT_BRAIN_URL = "https://ghost-healer-brain.onrender.com"
BUILTIN_API_KEY = "gh_sdk_public_8f4a2c9e1b7d3f6a0e5c8b2d4f7a1e9"


def ghost_dir() -> Path:
    return Path.home() / ".ghost"


def credentials_path() -> Path:
    return ghost_dir() / "credentials.json"


def load_global_credentials() -> Optional[Dict[str, Any]]:
    path = credentials_path()
    if not path.is_file():
        return None
    try:
        return json.loads(path.read_text(encoding="utf-8"))
    except (json.JSONDecodeError, OSError):
        return None


def save_global_credentials(
    api_key: str,
    brain_url: str = DEFAULT_BRAIN_URL,
    tenant_id: str = "default",
    project_id: str = "default",
    source: str = "user",
) -> Dict[str, Any]:
    ghost_dir().mkdir(parents=True, exist_ok=True)
    payload = {
        "api_key": api_key,
        "brain_url": brain_url,
        "tenant_id": tenant_id,
        "project_id": project_id,
        "source": source,
    }
    path = credentials_path()
    path.write_text(json.dumps(payload, indent=2), encoding="utf-8")
    try:
        path.chmod(0o600)
    except OSError:
        pass
    return payload


def ensure_builtin_credentials() -> Dict[str, Any]:
    existing = load_global_credentials()
    if existing:
        return existing
    return save_global_credentials(
        api_key=BUILTIN_API_KEY,
        brain_url=DEFAULT_BRAIN_URL,
        tenant_id="sdk",
        project_id="default",
        source="builtin",
    )


def apply_global_credentials() -> None:
    """Apply saved or built-in SDK credentials (does not override existing env)."""
    creds = load_global_credentials()
    if creds:
        if creds.get("api_key") and not os.environ.get("GHOST_API_KEY"):
            os.environ["GHOST_API_KEY"] = str(creds["api_key"])
        if creds.get("brain_url") and not os.environ.get("GHOST_BRAIN_URL"):
            os.environ["GHOST_BRAIN_URL"] = str(creds["brain_url"])
        if creds.get("tenant_id") and not os.environ.get("GHOST_TENANT_ID"):
            os.environ["GHOST_TENANT_ID"] = str(creds["tenant_id"])
        if creds.get("project_id") and not os.environ.get("GHOST_PROJECT_ID"):
            os.environ["GHOST_PROJECT_ID"] = str(creds["project_id"])

    if not os.environ.get("GHOST_API_KEY"):
        os.environ["GHOST_API_KEY"] = BUILTIN_API_KEY
    if not os.environ.get("GHOST_BRAIN_URL"):
        os.environ["GHOST_BRAIN_URL"] = DEFAULT_BRAIN_URL

File: sdk/ts/src/credentials.js
'use strict';

const fs = require('fs');
const path = require('path');
const os = require('os');

const DEFAULT_BRAIN_URL = 'https://ghost-healer-brain.onrender.com';

function getGhostDir() {
  return path.join(os.homedir(), '.ghost');
}

function getCredentialsPath() {
  return path.join(getGhostDir(), 'credentials.json');
}

function loadGlobalCredentials() {
  try {
    const p = getCredentialsPath();
    if (!fs.existsSync(p)) return null;
    return JSON.parse(fs.readFileSync(p, 'utf8'));
  } catch {
    return null;
  }
}

function saveGlobalCredentials(creds) {
  const dir = getGhostDir();
  fs.mkdirSync(dir, { recursive: true, mode: 0o700 });
  const payload = {
    api_key: creds.api_key || '',
    brain_url: creds.brain_url || DEFAULT_BRAIN_URL,
    tenant_id: creds.tenant_id || 'default',
    project_id: creds.project_id || 'default',
    saved_at: new Date().toISOString(),
  };
  fs.writeFileSync(getCredentialsPath(), JSON.stringify(payload, null, 2), { mode: 0o600 });
  return payload;
}

/** Apply ~/.ghost/credentials.json and optional enterprise URL to process.env */
function applyGlobalCredentials() {
  const creds = loadGlobalCredentials();
  if (creds) {
    if (creds.api_key && !process.env.GHOST_API_KEY) {
      process.env.GHOST_API_KEY = creds.api_key;
    }
    if (creds.brain_url && !process.env.GHOST_BRAIN_URL) {
      process.env.GHOST_BRAIN_URL = creds.brain_url;
    }
    if (creds.tenant_id && !process.env.GHOST_TENANT_ID) {
      process.env.GHOST_TENANT_ID = creds.tenant_id;
    }
    if (creds.project_id && !process.env.GHOST_PROJECT_ID) {
      process.env.GHOST_PROJECT_ID = creds.project_id;
    }
  }

  // Enterprise: IT hosts a JSON credentials file (internal HTTPS)
  const credUrl = process.env.GHOST_CREDENTIALS_URL;
  if (credUrl && !process.env.GHOST_API_KEY) {
    try {
      const https = require('https');
      const http = require('http');
      const client = credUrl.startsWith('https') ? https : http;
      // Sync fetch not available — skip in sync path; async handled in fetchCredentialsUrl
    } catch {
      /* optional */
    }
  }
}

function hasApiKey() {
  applyGlobalCredentials();
  return Boolean((process.env.GHOST_API_KEY || '').trim());
}

module.exports = {
  DEFAULT_BRAIN_URL,
  getGhostDir,
  getCredentialsPath,
  loadGlobalCredentials,
  saveGlobalCredentials,
  applyGlobalCredentials,
  hasApiKey,
};
